### PR TITLE
drivers: spi_engine: Added platform agnostic spi engine

### PR DESCRIPTION
**Pull request by scuciurean**
Adapted to the new no-OS structure.

v1->v2 :
              - Removed the CLA warnings
              - Minor style rework

Signed-off-by: Sergiu Cuciurean sergiu.cuciurean@analog.com

### CODE REVIEW

File: drivers/axi_core/spi_engine/spi_engine.c [hunk 1]
```diff
@@ -0,0 +1,943 @@
...
+		spi_eng_write(desc_extra,
+			      SPI_ENGINE_REG_CMD_FIFO, xfer->cmd_fifo[i]);
+	}
+
+	/* Write a number of tx_length WORDS on the SDO line */
+	for(i = 0; i < desc_extra->tx_length; i++) {
+		spi_eng_write(desc_extra,
+			      SPI_ENGINE_REG_SDO_DATA_FIFO, msg->tx_buf[i]);
+	}
+
```
Comment (mirceacaprioru) [inline drivers/axi_core/spi_engine/spi_engine.c]: nitpick: extra line

File: drivers/axi_core/spi_engine/spi_engine.c [hunk 2]
```diff
@@ -0,0 +1,943 @@
...
+#include "spi.h"
+#include "spi_engine.h"
+
+/******************************************************************************/
+/********************** Macros and Constants Definitions **********************/
+/******************************************************************************/
+
+#define ARRAY_SIZE(x)	(sizeof(x) / sizeof((x)[0]))
+
+static inline struct spi_init_param_extra *cast_to_extra_init(void* void_param)
```
Comment (mirceacaprioru) [inline drivers/axi_core/spi_engine/spi_engine.c]: do we actually need these wrappers ?

File: drivers/axi_core/spi_engine/spi_engine.c [hunk 3]
```diff
@@ -0,0 +1,943 @@
...
+						desc_extra->data_width));
+
+	/* Add a sync command to signal that the transfer has started */
+	spi_eng_program_add_cmd(xfer,
+				SPI_ENGINE_CMD_SYNC(SPI_ENGINE_SYNC_TRANSFER_BEGIN));
+
+	/* Add the rest of the commands */
+	for (i = 0; i < n; i++) {
+		spi_eng_add_user_cmd(desc, xfer, msg->spi_msg_cmds[i]);
+	}
```
Comment (mirceacaprioru) [inline drivers/axi_core/spi_engine/spi_engine.c]: these brackets can be removed. If you are doing this revision, a clean-up would not hurt.

File: drivers/axi_core/spi_engine/spi_engine.c [hunk 4]
```diff
@@ -0,0 +1,943 @@
...
+
+	xfer->cmd_fifo_len = 0;
+
+	spi_eng_compile_message(desc, msg, xfer);
+
+	/* Write the command fifo buffer */
+	for (i = 0; i < xfer->cmd_fifo_len; i++) {
+		spi_eng_write(desc_extra,
+			      SPI_ENGINE_REG_CMD_FIFO, xfer->cmd_fifo[i]);
+	}
```
Comment (mirceacaprioru) [inline drivers/axi_core/spi_engine/spi_engine.c]: same as above

File: drivers/axi_core/spi_engine/spi_engine.c (hunk 2 above)
Comment (scuciurean) [reply drivers/axi_core/spi_engine/spi_engine.c]: I've tried to adopt an approach similar to IIO:

`
static inline struct iio_dev *iio_priv_to_dev(void *priv)
{
	return (struct iio_dev *)((char *)priv -
				  ALIGN(sizeof(struct iio_dev), IIO_ALIGN));
}
`

**Pull request closed without merging**